src/v/datalake/catalog_schema_manager.cc

```diff
@@ -56,23 +61,28 @@ fill_field_ids(iceberg::struct_type& dest, const iceberg::struct_type& source) {
     for (auto& f : std::ranges::reverse_view(source.fields)) {
         source_stack.emplace_back(f.get());
     }
+    bool has_primitive_type_promotion{false};
     while (!source_stack.empty() && !dest_stack.empty()) {
         auto* dst = dest_stack.back();
         auto* src = source_stack.back();
-        if (
-          dst->name != src->name || dst->required != src->required
-          || dst->type.index() != src->type.index()) {
-            return fill_errc::mismatch;
+        if (auto compatibility = check_types(src->type, dst->type);
+            dst->name != src->name || dst->required != src->required
+            || compatibility.has_error()) {
+            return fill_errc::invalid_schema;
+        } else if (compatibility.value() == type_promoted::yes) {
+            has_primitive_type_promotion = true;
```

Comment:
Could we have just returned early here? Or is the intent that we still fill the field IDs when there is a promotion?

Reply:
Ah 🤦 yea makes sense. Thanks!